feat: add fullscreen button to controls widget

GregoryConrad opened the pull request:
## Description

Adds a fullscreen button to the controls widget.

Adds a dependency on `fullscreen_window`, but I could hand roll a solution if you didn't want to include another external dependency.

You can view this in action at https://rearch-presentation.web.app/ (hit the up arrow key to open the navigation drawer).

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

mkobuolys commented:
Hey @GregoryConrad, thanks for the initiative! I just checked the code, and here is my feedback:
1. `fullscreen_window` is not supported on all platforms (e.g. you get an exception when pressing the button on desktop platforms). So you need to cover the desktop full-screen cases explicitly or hide the button on non-supported platforms.
2. I am not a fan of the current position of the button. Before merging the code changes, I would try to play a bit with the navigation drawer and find a better placement.
3. Also, it would make sense to create a keyboard shortcut to toggle full-screen mode.

Let me know what you think. I think this could be a nice feature to add to the framework, just it need some work, still.

GregoryConrad commented:
Thanks for the feedback:

1. I added a platform check so the button will be hidden on unsupported platforms. This unfortunately requires a bit of a mess with conditional imports but I am not sure there is another way to do it without forking `fullscreen_window` and also contributing a `FullscreenWindow.isPlatformSupported()` sort of function.
2. No worries, feel free to make your changes to this PR/branch. I'm not tied to the button position!
3. This is where this feature can get a little funny, and I did think about this when I made the PR--there's no way to check if we are currently in fullscreen. So, I could add a keybinding, but the key will only ever be able to open fullscreen and not close it (and that behavior may be a bit unintuitive). Also, this particular key should likely be optional and default to null anyways, since it is only supported on particular platforms (and adding a new keybinding is a breaking change for applications that expect certain keyboard input). Especially since the behavior may need to change across platforms, I wasn't sure it'd be worth it to include a key that only sometimes works.

GregoryConrad commented:
Hi @mkobuolys ,

Would you mind giving this another review with my above comments in mind? Updated the PR to support the control panel you introduced in v0.9. Thanks.

GregoryConrad commented:
Odd, my branch is fine/correct locally but GitHub is throwing a fit here even after a force push. I'll see if I can fix it...

mkobuolys commented:
No worries, I will revisit this PR and check a few things (where to put the button, maybe add a shortcut, double-check if the 3rd-party dependency is really needed or maybe move the code to the core logic and maintain it here, etc.).

GregoryConrad commented:
I took at the library‘s source code again—I’m frankly not sure we need to add the external dependency. Considering we already need to have conditional includes, I think I’ll just toss the relevant line or two of code to enable full screen in this PR later today.

GregoryConrad commented:
Overlooked the source code of that package a bit a few hours ago. Here's an update"

iOS, Android are one liners:
`SystemChrome.setEnabledSystemUIMode(SystemUiMode.immersiveSticky);`

Web is also trivial:
`html.window.document.documentElement?.requestFullscreen();`

Windows, not so much. We may be able to do it with a dependency on the `win32` package: https://pub.dev/packages/win32

macOS would likely require something similar to windows here.

Let me know what you think. If it weren't for the complexity of windows, I'd say drop the dependency. But since windows is a little bit trickier, it may just be worth keeping the dependency we have instead of swapping it for a new one.

GregoryConrad commented:
@mkobuolys And if we stick with `fullscreen_window`, this PR is ready to go as-is.

mkobuolys commented:
The only concern I have is that I would like to be able to detect whether the current view is fullscreen or not. It should be possible to detect this on the Web (check how Google Slides works), I am almost sure that it should be possible to do that on other platforms as well. Also, I am not entirely sure if this feature is needed on Android/iOS (I think it's needed, let's say, for iPad, when you can run your app on only one half of the screen, but that's more of an edge case). So I would be up for starting small and only shipping the full feature, let's say, for Web only. That's mostly why I am thinking about removing the dependency and only bringing the code I need.

GregoryConrad commented:
@mkobuolys sounds good! I’ll drop the dependency and implement the toggle-like functionality for web (and android/iOS if I can get it to work).

GregoryConrad commented:
Updated PR to drop `fullscreen_window` dependency and add just web support. Couldn't find a readily-available Flutter API on Android/iOS to check if the application is currently in fullscreen without dropping to native code, so decided to leave those platforms unimplemented in this PR.

GregoryConrad commented:
@mkobuolys Any update here? PR should be ready to go with your suggestions (and I also checked the latest changes you pushed to `main` to see if those were applicable but I don't think so since fullscreen is only available on web atm).

mkobuolys commented on lib/src/controls/flutter_deck_controls.dart:
```diff
@@ -246,6 +248,19 @@ class _MarkerButton extends StatelessWidget {
+    return MenuItemButton(
+      leadingIcon: const Icon(Icons.fullscreen_rounded),
+      onPressed: isInFullscreen() ? leaveFullscreen : enterFullscreen,
+      child: Text(isInFullscreen() ? 'Leave fullscreen' : 'Enter fullscreen'),
```
```suggestion
      child: Text(isInFullscreen() ? 'Exit full screen' : 'Enter full screen'),
```

mkobuolys commented on lib/src/controls/flutter_deck_controls.dart:
```diff
@@ -246,6 +248,19 @@ class _MarkerButton extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return MenuItemButton(
+      leadingIcon: const Icon(Icons.fullscreen_rounded),
```
```suggestion
      leadingIcon: Icon(isInFullscreen() ? Icons.fullscreen_exit_rounded : Icons.fullscreen_rounded),
```

mkobuolys reviewed (changes requested):
This is what I would do differently (based on architectural preferences and keeping it consistent across the current codebase):

**Basic refactoring**
Move `fullscreen` logic under the `controls` directory;

**A matter of taste, but mostly for the sake of consistency**
I am not a fan of public methods and you won't probably find them across the whole project. To me, instead of exposing all the fullscreen methods, I would create something like `FlutterDeckFullscreenManager` and put the methods inside the class. Even better, I would create some base class or interface, like `FlutterDeckFullscreenManagerBase` that contains all the stub methods (`canFullscreen() => false` and all the `UnsupportedError` implementations). Then, for the `io.dart` and `stub.dart` I would do something like `class FlutterDeckFullscreenManager extends FlutterDeckFullscreenManagerBase {}` - that's the whole implementation needed for this export. Only the `web.dart` extends the base class **AND OVERRIDES** it with the needed logic. To me, this is a more OOP-ish way to encapsulate the logic and later use it wherever is needed. Now, where to put it... (\**cough\**, see the next section). In fact, this is basically how federated plugins are implemented in Flutter.

**A bit more refactoring**
I do not like the way how fullscreen methods are used. The main idea is to keep all the controls-related logic under `FlutterDeckControlsNotifier` so that you use it as the main entry point to control the deck. Thus, inside the `flutter_deck_app.dart` initialise the newly created `FlutterDeckFullscreenManager` and pass it to the `FlutterDeckControlsNotifier`. Inside the `FlutterDeckControlsNotifier`, expose the methods that would proxy to whatever logic was exported. I think these methods should not call the `notifyListeners()`, but please double-check it (e.g. fullscreen button icon is not updated or something is not in sync). This will allow us to have all the controls in a single place, and maybe it makes sense to extend the logic later on. E.g. handle keyboard shortcuts more easily, or disable fullscreen action when other actions, like marker, are enabled.

mkobuolys commented:
@GregoryConrad Sorry for the late answer (vacation + Holidays). I think that's enough input for now. Let me know if you have any questions ✌️

GregoryConrad commented:
Thanks for the suggestions! They should be all implemented now.

> I am not a fan of public methods and you won't probably find them across the whole project. To me, instead of exposing all the fullscreen methods, I would create something like FlutterDeckFullscreenManager and put the methods inside the class. 

Refactored the PR to use this approach and more OOP per your suggestions. I'm getting a little too used to Rust where you don't have to make a "wrapper" class for a collection of functions and instead can just toss your code in a new module and have everything be well organized; that habit was bleeding over. I also normally don't even touch OOP with a 10 foot poll (a few meter poll for non-USA friends), so sorry about the differences with the rest of the code base! I refactored it and also made the `FlutterDeckControlsNotifier` implement the `FlutterDeckControlsManagerBase` to ensure that (1) all methods in base will always be implemented (won't accidentally forget one) and (2) doc comments are automatically brought over from the base. I also toyed with the idea of a `base class FlutterDeckManagerBase`, but this prevents `implements FlutterDeckManagerBase`. Also tried this as an `interface class` with default members, but then in `FlutterDeckManager` it is not possible to use the default implementations. Thus, I just stuck with a boring old `abstract class` to get more flexibility.

> Sorry for the late answer (vacation + Holidays). I think that's enough input for now. Let me know if you have any questions ✌️

No worries! I am also on vacation but haven't had anything to do at night/morning 😝. Happy holidays!

mkobuolys reviewed (approved):
Looking good, thanks for your contribution and happy holidays! I will add the release notes and update the documentation myself with another feature (spoiler alert, autoplay) ✌️
